### ISSUE TITLE
No generated action stubs for async actions returning a derived ActionResult type

### ISSUE THREAD

**Issue by DoctorVanGogh** [T4MVC/T4MVC#57]
I've run into a nasty surprise trying out T4MVC: There will be _no_ generated action method stubs for actions that are declared as both async and a derived action type.

While

``` csharp
public async Task<ActionResult> FooActionResultAsync() {     
    return Content("Foo");
}
```

works fine, the following will not:

``` csharp
public async Task<JsonResult> FooJsonResultAsync() {
    return Json("Foo");
}

public async Task<PartialViewResult> FooPartialResultAsync() {
    return PartialView("FooView");
}
```

Basically if an async action method does not use `ActionResult` then the stub does not get generated.
Surprisingly, the `T4MVC_System_Web_Mvc_[Derived]Result` types still get generated.

I think the issue stems from a literal comparison in the T4 of (in multiple places)

```
method.Type.CodeType.FullName =="System.Threading.Tasks.Task<System.Web.Mvc.ActionResult>
```

I think a check like

``` csharp
var t = method.Type.CodeType;

if (t.IsConstructedGenericType
    && t.GetGenericTypeDefinition() == typeof (Task<>)
    && (t.GetGenericArguments()[0] == typeof (ActionResult)
        || t.GetGenericArguments()[0].IsSubclassOf(typeof (ActionResult)))
```

would be more appropriate here. Of course, the related 

``` csharp
 return System.Threading.Tasks.Task.FromResult(callInfo as ActionResult);
```

statements would have to be updated as well to output the actual action result type.

No idea, if there is a deeper reason, why those async methods are not supported, but it'd sure be nice, if they were ;)

**Comment by davidebbo** [comment]
Not sure. It's probably just something that hadn't come up. Please send PR if you have possible fix that doesn't regress existing scenarios.

**Comment by DoctorVanGogh** [comment]
I've added a PR - now I remember why I dislike writing T4s. The type system is severely limited.

I _think_ I got all the overloads that would be needed for the T4Extensions, but I might have missed one or two.
Maybe the original versions _could_ be removed, but I don't know about binary compatibility of external consumers if you simply replace all `Foo(... Task<ActionResult> ...)` methods with `Foo<TActionResult>(... Task<TActionResult> ...) where TActionResult : ActionResult` versions.
